feat: introduce Mitosis execution node with EVM integration

riemannulus opened the pull request:
- Added `mi-reth` binary for Mitosis execution node.
- Implemented Mitosis-specific EVM configuration and execution logic.
- Integrated `ress` subprotocol support for enhanced functionality.
- Updated workspace to include new Mitosis crates and dependencies.
- Enhanced Makefile with build and install commands for Mitosis.
- Added necessary CLI and library files for Mitosis operations.

riemannulus committed:
feat: introduce Mitosis execution node with EVM integration

- Added `mi-reth` binary for Mitosis execution node.
- Implemented Mitosis-specific EVM configuration and execution logic.
- Integrated `ress` subprotocol support for enhanced functionality.
- Updated workspace to include new Mitosis crates and dependencies.
- Enhanced Makefile with build and install commands for Mitosis.
- Added necessary CLI and library files for Mitosis operations.

taeguk commented on bin/mitosis/Cargo.toml:
```diff
@@ -0,0 +1,139 @@
+[package]
+name = "mi-reth"
```
cool name 👍

taeguk commented on crates/mitosis/evm/src/executor.rs:
```diff
@@ -0,0 +1,120 @@
+
+        self.inner.apply_pre_execution_changes()?;
+        
+        // Deploy Multicall3 contract using system transaction at block 150
```
```suggestion
        // Deploy Multicall3 contract using system transaction at a specific height
```

taeguk commented on crates/mitosis/evm/src/executor.rs:
```diff
@@ -0,0 +1,120 @@
+    type Evm = Inner::Evm;
+
+    fn apply_pre_execution_changes(&mut self) -> Result<(), BlockExecutionError> {
+        // First apply the inner pre-execution changes
+
+        self.inner.apply_pre_execution_changes()?;
```
```suggestion
        // First apply the inner pre-execution changes
        self.inner.apply_pre_execution_changes()?;
```

taeguk commented on crates/mitosis/evm/src/executor.rs:
```diff
@@ -0,0 +1,120 @@
+        self.inner.apply_pre_execution_changes()?;
+        
+        // Deploy Multicall3 contract using system transaction at block 150
+        apply_multicall3_deployment(self.inner.evm_mut())?;
```
I think we should apply this only for mainnet. We should not run this logic when it is not a mainnet.

riemannulus replied:
https://github.com/mitosis-org/reth/pull/1/files#diff-bcfae8e1e0afb3cc7090746b8ecf7d35fd2d3aa08297632b8a2f4c2de7d78211R28-R40

I fix it. we should change this feature to real hardfork function when we ready.

taeguk commented on crates/mitosis/evm/src/system_calls.rs:
```diff
@@ -0,0 +1,84 @@
+    CONTRACT_REPLACEMENT_BLOCK, MULTICALL3_ADDRESS, get_multicall3_bytecode
+};
+
+/// Deploys the Multicall3 contract at block 150 by directly modifying the state.
```
```suggestion
/// Deploys the Multicall3 contract at a specific height by directly modifying the state.
```

taeguk commented on crates/mitosis/evm/src/system_calls.rs:
```diff
@@ -0,0 +1,84 @@
+    let bytecode = get_multicall3_bytecode();
+    
+    // Create the account info with the bytecode
+    let account_info = AccountInfo {
+        balance: U256::ZERO,
+        nonce: 1, // Contract nonce (1 indicates it's a contract)
+        code_hash: keccak256(&bytecode),
+        code: Some(Bytecode::new_raw(bytecode)),
+    };
```
```suggestion
    let account_info = AccountInfo {
        balance: _existing_account.map(|acc| acc.balance).unwrap_or(U256::ZERO),
        nonce: 1, // Contract nonce (1 indicates it's a contract)
        code_hash: keccak256(&bytecode),
        code: Some(Bytecode::new_raw(bytecode)),
    };
```

riemannulus replied:
fix it

taeguk commented on crates/mitosis/primitives/src/lib.rs:
```diff
@@ -0,0 +1,22 @@
+]);
+
+/// The block number at which the contract replacement should occur
+pub const CONTRACT_REPLACEMENT_BLOCK: u64 = 10;
```
```suggestion
pub const MULTICALL3_REPLACEMENT_BLOCK: u64 = 10;
```

taeguk commented on crates/mitosis/primitives/src/lib.rs:
```diff
@@ -0,0 +1,22 @@
+
+use alloy_primitives::{Address, Bytes};
+
+/// The Multicall3 contract address that will be deployed at block 150
```
```suggestion
/// The Multicall3 contract address that will be deployed at a specific height
```

riemannulus committed:
refactor: update Multicall3 deployment logic and related constants

- Refined comments for clarity regarding the deployment of the Multicall3 contract.
- Introduced HARDFORK_CHAIN_ID constant to specify the chain ID for Multicall3 deployment.
- Removed outdated block number checks and adjusted logic to ensure deployment occurs at the correct chain ID and block number.
- Cleaned up unnecessary code and comments in the executor and system calls files.

riemannulus committed:
fix: update Multicall3 replacement block number

- Changed the MULTICALL3_REPLACEMENT_BLOCK constant from 10 to 786000 to reflect the correct block number for contract replacement.
- Ensured consistency with the latest deployment logic adjustments.
